src/gafaelfawr/services/token.py

```diff
@@ -0,0 +1,473 @@
+            The name of the token.
+        scopes : List[`str`] or `None`
+            The scopes of the token.
+        expires : `datetime` or `None`
+            When the token should expire.  If not given, defaults to the
+            expiration of the authentication token taken from ``data``.
+        no_expire : `bool`
+            If set, the token should not expire.  This is a separate parameter
+            because passing `None` to ``expires`` is ambiguous.
+
+        Returns
+        -------
+        token : `gafaelfawr.models.token.Token`
+            The newly-created token.
+
+        Raises
+        ------
+        gafaelfawr.exceptions.PermissionDeniedError
+            If the given username didn't match the user information in the
+            authentication token.
+        """
+        if username != auth_data.username:
+            msg = "Cannot create tokens for another user"
+            raise PermissionDeniedError(msg)
+
+        token = Token()
+        created = datetime.now(tz=timezone.utc).replace(microsecond=0)
+        if no_expire:
+            expires = None
+        elif not expires:
+            expires = auth_data.expires
+        data = TokenData(
+            token=token,
+            username=auth_data.username,
+            token_type=TokenType.user,
+            scopes=sorted(scopes) if scopes else [],
+            created=created,
+            expires=expires,
+            name=auth_data.name,
+            uid=auth_data.uid,
+            groups=auth_data.groups,
+        )
+        await self._token_redis_store.store_data(data)
+        with self._transaction_manager.transaction():
+            self._token_db_store.add(data, token_name=token_name)
+        self._logger.info(
+            "Created new user token",
+            key=token.key,
+            token_name=token_name,
+            token_scope=",".join(data.scopes),
+        )
+        return token
+
+    async def delete_token(
+        self, key: str, auth_data: TokenData, username: Optional[str] = None
+    ) -> bool:
+        """Delete a token.
+
+        Parameters
+        ----------
+        key : `str`
+            The key of the token to delete.
+        auth_data : `gafaelfawr.models.token.TokenData`
+            The token data for the authentication token of the user deleting
+            the token.
+        username : `str`, optional
+            If given, constrain deletions to tokens owned by the given user.
+
+        Returns
+        -------
+        success : `bool`
+            Whether the token was found and deleted.
+        """
+        info = self.get_info(key, username)
+        if not info:
+            return False
+        if info.username != auth_data.username:
+            msg = f"Token owned by {info.username}, not {auth_data.username}"
+            self._logger.warning("Permission denied", error=msg)
+            raise PermissionDeniedError(msg)
+        await self._token_redis_store.delete(key)
```

Comment:
Are there any potential race conditions by having this removed from Redis before Postgres?  (Or vice versa for creation, where it appears in Redis before Postgres.)

Reply:
Redis is canonical and PostgreSQL is secondary, so it's harmless for it to appear in PostgreSQL but not Redis (and will eventually be cleaned up by the as-yet-to-be-written housekeeping job). Good point on creation; I've now reversed the order so that the token is created in PostgreSQL first and then in Redis, on the principle that the PostgreSQL orphans are the harmless ones.